modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectEntryLocalServiceImpl.java

```diff
@@ -1628,6 +1632,119 @@ public ObjectEntry updateObjectEntry(
+
+		Connection connection = _currentConnection.getConnection(
+			objectEntryPersistence.getDataSource());
+
+		String sql1 =
+			"update ObjectEntry set rootObjectEntryId = ? where " +
+				"objectEntryId = ?";
+		String sql2 =
+			"update ObjectEntry set rootObjectEntryId = ? where " +
+				"rootObjectEntryId = ?";
+
+		try (PreparedStatement preparedStatement1 = connection.prepareStatement(
+				sql1);
+			PreparedStatement preparedStatement2 = connection.prepareStatement(
+				sql2)) {
+
+			_performActionableDynamicQuery(
+				objectDefinition1.getObjectDefinitionId(), false,
+				(ObjectEntry objectEntry) -> {
+					try {
+						long rootObjectEntryId =
+							objectEntry.getRootObjectEntryId();
+
+						if (rootObjectEntryId == 0) {
+							rootObjectEntryId = objectEntry.getObjectEntryId();
+
+							preparedStatement1.setLong(1, rootObjectEntryId);
+
+							preparedStatement1.setLong(
+								2, objectEntry.getObjectEntryId());
+
+							preparedStatement1.addBatch();
+						}
+
+						for (ObjectEntry relatedObjectEntry :
+								getOneToManyObjectEntries(
+									0,
+									objectRelationship.
+										getObjectRelationshipId(),
+									objectEntry.getObjectEntryId(), true, null,
+									QueryUtil.ALL_POS, QueryUtil.ALL_POS)) {
+
+							if (objectDefinition2PreviousRootObjectDefinitionId ==
+									0) {
+
+								preparedStatement1.setLong(
+									1, rootObjectEntryId);
+								preparedStatement1.setLong(
+									2, relatedObjectEntry.getObjectEntryId());
+
+								preparedStatement1.addBatch();
+
+								continue;
+							}
+
+							preparedStatement2.setLong(1, rootObjectEntryId);
+							preparedStatement2.setLong(
+								2, relatedObjectEntry.getObjectEntryId());
+
+							preparedStatement2.addBatch();
+						}
+					}
+					catch (SQLException sqlException) {
+						throw new SystemException(sqlException);
+					}
+				});
+
+			preparedStatement1.executeBatch();
+
+			if (objectDefinition2PreviousRootObjectDefinitionId > 0) {
+				preparedStatement2.executeBatch();
+			}
+		}
+		catch (SQLException sqlException) {
+			throw new SystemException(sqlException);
+		}
+
+		if (objectDefinition2PreviousRootObjectDefinitionId == 0) {
+			return;
+		}
+
+		TreeFactory treeFactory = _treeFactorySnapshot.get();
+
+		Tree objectDefinitionTree = treeFactory.createObjectDefinitionTree(
+			objectDefinition1.getObjectDefinitionId(),
+			_objectDefinitionPersistence::findByPrimaryKey);
+
+		Iterator<Node> iterator = objectDefinitionTree.iterator();
+
+		while (iterator.hasNext()) {
+			Node node = iterator.next();
+
+			ObjectDefinition objectDefinition =
+				_objectDefinitionPersistence.findByPrimaryKey(
+					node.getPrimaryKey());
+
+			Indexer<ObjectEntry> indexer = IndexerRegistryUtil.getIndexer(
+				objectDefinition.getClassName());
+
+			_performActionableDynamicQuery(
+				objectDefinition1.getObjectDefinitionId(), true,
+				(ObjectEntry objectEntry) -> indexer.reindex(objectEntry));
```

Comment:
_performActionableDynamicQuery(
	**objectDefinition1**.getObjectDefinitionId(), true,
	(ObjectEntry objectEntry) -> indexer.reindex(objectEntry));

Double-checking: is this supposed to always be the **objectDefinition1** even when the `iterator.next()` changes to another object definition of the objectDefinitionTree?

Reply:
That was inattention, I forgot to update it from a previous logic. 

Also, while analyzing this part again, I noticed we do not always have to reindex `objectDefinition1`, so I added a conditional to avoid doing so unnecessarily on this [commit](https://github.com/liferay-objects/liferay-portal/pull/3682/commits/bd10fb0fb4913ffa716e70c73ae7f9210ef97c51).